File: src/lib/logger.ts
/**
 * Module de journalisation simplifié pour le frontend
 */

// Types des niveaux de journalisation
export type LogLevel = 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL';
export type ExportFormat = 'json' | 'csv' | 'txt';

// Interface pour les entrées de journal
export interface LogEntry {
  timestamp: string;
  level: LogLevel;
  module: string;
  message: string;
  data?: Record<string, any>;
}

/**
 * Classe Logger pour gérer les journaux dans l'application
 */
class Logger {
  private logs: LogEntry[] = [];
  private maxEntries: number;
  private level: LogLevel;
  private levelPriority: Record<LogLevel, number> = {
    'DEBUG': 0,
    'INFO': 1,
    'WARNING': 2,
    'ERROR': 3,
    'CRITICAL': 4
  };

  /**
   * Initialise le logger
   */
  constructor(maxEntries: number = 10000, level: LogLevel = 'INFO') {
    this.maxEntries = maxEntries;
    this.level = level;
    
    // Charger les logs depuis le localStorage s'ils existent
    this.loadFromStorage();
  }

  /**
   * Détermine si un message doit être journalisé selon son niveau
   */
  private shouldLog(level: LogLevel): boolean {
    return this.levelPriority[level] >= this.levelPriority[this.level];
  }

  /**
   * Sauvegarde les logs dans le localStorage
   */
  private saveToStorage(): void {
    try {
      // Ne garder que les 1000 dernières entrées pour le localStorage
      const logsToSave = this.logs.slice(-1000);
      localStorage.setItem('appLogs', JSON.stringify(logsToSave));
    } catch (error) {
      console.error('Erreur lors de la sauvegarde des logs:', error);
    }
  }

  /**
   * Charge les logs depuis le localStorage
   */
  private loadFromStorage(): void {
    try {
      const savedLogs = localStorage.getItem('appLogs');
      if (savedLogs) {
        this.logs = JSON.parse(savedLogs);
      }
    } catch (error) {
      console.error('Erreur lors du chargement des logs:', error);
    }
  }

  /**
   * Ajoute une entrée au journal
   */
  log(message: string, level: LogLevel = 'INFO', module: string = '', additionalData?: Record<string, any>): void {
    if (!this.shouldLog(level)) {
      return;
    }

    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      module,
      message
    };

    if (additionalData) {
      entry.data = additionalData;
    }

    this.logs.push(entry);

    // Limiter la taille du journal
    if (this.logs.length > this.maxEntries) {
      this.logs = this.logs.slice(-this.maxEntries);
    }

    // Sauvegarder les logs dans le localStorage
    this.saveToStorage();

    // En mode développement, afficher également dans la console
    const consoleMethod = this.getConsoleMethod(level);
    consoleMethod(`[${level}] ${module ? `${module}: ` : ''}${message}`);
  }

  /**
   * Obtient la méthode console appropriée pour le niveau de log
   */
  private getConsoleMethod(level: LogLevel): (...data: any[]) => void {
    switch (level) {
      case 'DEBUG': return console.debug;
      case 'INFO': return console.info;
      case 'WARNING': return console.warn;
      case 'ERROR': 
      case 'CRITICAL': return console.error;
      default: return console.log;
    }
  }

  /**
   * Log de niveau DEBUG
   */
  debug(message: string, module: string = '', additionalData?: Record<string, any>): void {
    this.log(message, 'DEBUG', module, additionalData);
  }

  /**
   * Log de niveau INFO
   */
  info(message: string, module: string = '', additionalData?: Record<string, any>): void {
    this.log(message, 'INFO', module, additionalData);
  }

  /**
   * Log de niveau WARNING
   */
  warning(message: string, module: string = '', additionalData?: Record<string, any>): void {
    this.log(message, 'WARNING', module, additionalData);
  }

  /**
   * Log de niveau ERROR
   */
  error(message: string, module: string = '', additionalData?: Record<string, any>): void {
    this.log(message, 'ERROR', module, additionalData);
  }

  /**
   * Log de niveau CRITICAL
   */
  critical(message: string, module: string = '', additionalData?: Record<string, any>): void {
    this.log(message, 'CRITICAL', module, additionalData);
  }

  /**
   * Exporte les journaux au format JSON
   */
  exportJson(): string {
    return JSON.stringify(this.logs, null, 2);
  }

  /**
   * Exporte les journaux au format CSV
   */
  exportCsv(): string {
    const headers = ['timestamp', 'level', 'module', 'message'];
    const headerRow = headers.join(',');
    
    const rows = this.logs.map(entry => {
      return [
        entry.timestamp,
        entry.level,
        `"${(entry.module || '').replace(/"/g, '""')}"`, // Échapper les guillemets
        `"${entry.message.replace(/"/g, '""')}"` // Échapper les guillemets
      ].join(',');
    });
    
    return [headerRow, ...rows].join('\n');
  }

  /**
   * Exporte les journaux au format texte
   */
  exportTxt(): string {
    return this.logs.map(entry => {
      return `[${entry.timestamp}] ${entry.level} - ${entry.module || ''}: ${entry.message}`;
    }).join('\n');
  }

  /**
   * Exporte les journaux dans le format spécifié
   */
  export(format: ExportFormat = 'json'): string {
    switch (format) {
      case 'json': return this.exportJson();
      case 'csv': return this.exportCsv();
      case 'txt': return this.exportTxt();
      default: throw new Error(`Format d'exportation non supporté: ${format}`);
    }
  }

  /**
   * Télécharge les journaux dans un fichier
   */
  downloadLogs(format: ExportFormat = 'json', filename?: string): void {
    // Générer le contenu à télécharger
    const content = this.export(format);
    
    // Déterminer le type MIME
    let mimeType: string;
    let extension: string;
    
    switch (format) {
      case 'json': 
        mimeType = 'application/json';
        extension = 'json';
        break;
      case 'csv': 
        mimeType = 'text/csv';
        extension = 'csv';
        break;
      case 'txt':
      default: 
        mimeType = 'text/plain';
        extension = 'txt';
        break;
    }
    
    // Générer un nom de fichier par défaut si nécessaire
    if (!filename) {
      const timestamp = new Date().toISOString().replace(/[-:]/g, '').split('.')[0].replace('T', '_');
      filename = `mabourse_logs_${timestamp}.${extension}`;
    }
    
    // Créer un blob et un lien de téléchargement
    const blob = new Blob([content], { type: mimeType });
    const url = URL.createObjectURL(blob);
    
    const a = document.createElement('a');
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    
    // Nettoyage
    setTimeout(() => {
      document.body.removeChild(a);
      URL.revokeObjectURL(url);
    }, 100);
  }

  /**
   * Efface toutes les entrées du journal
   */
  clear(): void {
    this.logs = [];
    localStorage.removeItem('appLogs');
  }

  /**
   * Retourne toutes les entrées du journal
   */
  getLogs(): LogEntry[] {
    return [...this.logs];
  }

  /**
   * Filtre les logs par niveau
   */
  filterByLevel(level: LogLevel): LogEntry[] {
    return this.logs.filter(entry => entry.level === level);
  }

  /**
   * Filtre les logs par module
   */
  filterByModule(module: string): LogEntry[] {
    return this.logs.filter(entry => entry.module === module);
  }
}

// Instance singleton du logger
const _logger = new Logger();

// Exporter les fonctions d'accès global
export function getLogger(): Logger {
  return _logger;
}

export function log(message: string, level: LogLevel = 'INFO', module: string = '', additionalData?: Record<string, any>): void {
  _logger.log(message, level, module, additionalData);
}

export function debug(message: string, module: string = '', additionalData?: Record<string, any>): void {
  _logger.debug(message, module, additionalData);
}

export function info(message: string, module: string = '', additionalData?: Record<string, any>): void {
  _logger.info(message, module, additionalData);
}

export function warning(message: string, module: string = '', additionalData?: Record<string, any>): void {
  _logger.warning(message, module, additionalData);
}

export function error(message: string, module: string = '', additionalData?: Record<string, any>): void {
  _logger.error(message, module, additionalData);
}

export function critical(message: string, module: string = '', additionalData?: Record<string, any>): void {
  _logger.critical(message, module, additionalData);
}

export function downloadLogs(format: ExportFormat = 'json', filename?: string): void {
  _logger.downloadLogs(format, filename);
}

export function clearLogs(): void {
  _logger.clear();
}

export default {
  getLogger,
  log,
  debug,
  info,
  warning,
  error,
  critical,
  downloadLogs,
  clearLogs
};

File: logging/logger.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Module de journalisation pour Ma Bourse
Permet de centraliser tous les messages de log et de les exporter
"""

import os
import json
import datetime
import csv
from typing import List, Dict, Any, Optional, Literal

# Niveaux de journalisation
LogLevel = Literal["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]

class Logger:
    """Gestionnaire de journalisation pour Ma Bourse"""
    
    def __init__(self, max_entries: int = 10000, level: LogLevel = "INFO"):
        """
        Initialise le logger
        
        Args:
            max_entries: Nombre maximum d'entrées à conserver en mémoire
            level: Niveau minimum de journalisation
        """
        self.logs: List[Dict[str, Any]] = []
        self.max_entries = max_entries
        self.level = level
        self.level_priority = {
            "DEBUG": 0,
            "INFO": 1,
            "WARNING": 2,
            "ERROR": 3,
            "CRITICAL": 4
        }
        
        # Créer un dossier pour les journaux exportés s'il n'existe pas
        self.log_dir = os.path.join(os.path.dirname(os.path.dirname(__file__)), "exported_logs")
        if not os.path.exists(self.log_dir):
            os.makedirs(self.log_dir)
    
    def _should_log(self, level: LogLevel) -> bool:
        """Détermine si un message doit être journalisé selon son niveau"""
        return self.level_priority[level] >= self.level_priority[self.level]
    
    def log(self, message: str, level: LogLevel = "INFO", module: str = "", 
            additional_data: Optional[Dict[str, Any]] = None) -> None:
        """
        Ajoute une entrée au journal
        
        Args:
            message: Le message à journaliser
            level: Le niveau de journalisation
            module: Le module source du message
            additional_data: Données supplémentaires à inclure
        """
        if not self._should_log(level):
            return
            
        entry = {
            "timestamp": datetime.datetime.now().isoformat(),
            "level": level,
            "module": module,
            "message": message
        }
        
        if additional_data:
            entry["data"] = additional_data
            
        self.logs.append(entry)
        
        # Limiter la taille du journal
        if len(self.logs) > self.max_entries:
            self.logs = self.logs[-self.max_entries:]
    
    def debug(self, message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
        """Log de niveau DEBUG"""
        self.log(message, "DEBUG", module, additional_data)
    
    def info(self, message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
        """Log de niveau INFO"""
        self.log(message, "INFO", module, additional_data)
    
    def warning(self, message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
        """Log de niveau WARNING"""
        self.log(message, "WARNING", module, additional_data)
    
    def error(self, message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
        """Log de niveau ERROR"""
        self.log(message, "ERROR", module, additional_data)
    
    def critical(self, message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
        """Log de niveau CRITICAL"""
        self.log(message, "CRITICAL", module, additional_data)
    
    def export_json(self, filename: Optional[str] = None) -> str:
        """
        Exporte les journaux au format JSON
        
        Args:
            filename: Nom du fichier d'export (généré automatiquement si None)
            
        Returns:
            Chemin du fichier exporté
        """
        if not filename:
            timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"mabourse_logs_{timestamp}.json"
        
        filepath = os.path.join(self.log_dir, filename)
        
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(self.logs, f, ensure_ascii=False, indent=2)
        
        return filepath
    
    def export_csv(self, filename: Optional[str] = None) -> str:
        """
        Exporte les journaux au format CSV
        
        Args:
            filename: Nom du fichier d'export (généré automatiquement si None)
            
        Returns:
            Chemin du fichier exporté
        """
        if not filename:
            timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"mabourse_logs_{timestamp}.csv"
        
        filepath = os.path.join(self.log_dir, filename)
        
        with open(filepath, 'w', newline='', encoding='utf-8') as f:
            fieldnames = ["timestamp", "level", "module", "message"]
            writer = csv.DictWriter(f, fieldnames=fieldnames)
            
            writer.writeheader()
            for entry in self.logs:
                # Filtrer uniquement les champs standard pour le CSV
                row = {field: entry.get(field, "") for field in fieldnames}
                writer.writerow(row)
        
        return filepath
    
    def export_txt(self, filename: Optional[str] = None) -> str:
        """
        Exporte les journaux au format texte
        
        Args:
            filename: Nom du fichier d'export (généré automatiquement si None)
            
        Returns:
            Chemin du fichier exporté
        """
        if not filename:
            timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"mabourse_logs_{timestamp}.txt"
        
        filepath = os.path.join(self.log_dir, filename)
        
        with open(filepath, 'w', encoding='utf-8') as f:
            for entry in self.logs:
                timestamp = entry["timestamp"]
                level = entry["level"]
                module = entry["module"]
                message = entry["message"]
                
                log_line = f"[{timestamp}] {level} - {module}: {message}"
                f.write(log_line + "\n")
        
        return filepath
    
    def export(self, format: Literal["json", "csv", "txt"] = "json", filename: Optional[str] = None) -> str:
        """
        Exporte les journaux dans le format spécifié
        
        Args:
            format: Format d'exportation ("json", "csv", "txt")
            filename: Nom du fichier d'export
            
        Returns:
            Chemin du fichier exporté
        """
        if format == "json":
            return self.export_json(filename)
        elif format == "csv":
            return self.export_csv(filename)
        elif format == "txt":
            return self.export_txt(filename)
        else:
            raise ValueError(f"Format d'exportation non supporté: {format}")
    
    def clear(self) -> None:
        """Efface toutes les entrées du journal"""
        self.logs = []

# Instance singleton du logger
_logger = Logger()

# Fonctions d'accès global au logger
def get_logger() -> Logger:
    """Retourne l'instance singleton du logger"""
    return _logger

def log(message: str, level: LogLevel = "INFO", module: str = "", 
        additional_data: Optional[Dict[str, Any]] = None) -> None:
    """Fonction globale pour journaliser un message"""
    _logger.log(message, level, module, additional_data)

def debug(message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
    """Fonction globale pour un log de niveau DEBUG"""
    _logger.debug(message, module, additional_data)

def info(message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
    """Fonction globale pour un log de niveau INFO"""
    _logger.info(message, module, additional_data)

def warning(message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
    """Fonction globale pour un log de niveau WARNING"""
    _logger.warning(message, module, additional_data)

def error(message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
    """Fonction globale pour un log de niveau ERROR"""
    _logger.error(message, module, additional_data)

def critical(message: str, module: str = "", additional_data: Optional[Dict[str, Any]] = None) -> None:
    """Fonction globale pour un log de niveau CRITICAL"""
    _logger.critical(message, module, additional_data)

def export_logs(format: Literal["json", "csv", "txt"] = "json", filename: Optional[str] = None) -> str:
    """Fonction globale pour exporter les journaux"""
    return _logger.export(format, filename)

def clear_logs() -> None:
    """Fonction globale pour effacer les journaux"""
    _logger.clear()
